### ISSUE TITLE
Molecule

### ISSUE THREAD

**Issue by fjr619** [mr3y-the-programmer/Podcaster#81]
Why arw you using molecule? What is the benefit? Cause i still confuse on that library, and by see your project that using molecule, still make me confuse, cause i still can achieve that in viewmodel without molecule too

**Comment by mr3y-the-programmer** [comment]
Of course you can, To better understand the benefit of using molecule in this project, you'll need to understand what is the requirement/problem that led me to use it in the first place.

In most screens of the app, we need an observable state holder that is:

1. Derived from many sources.
2. Updates state accordingly whenever any of those sources change.

That being said, we need a `StateFlow` of a screen/component state that'll continuously re-emit updates whenever state changes to notify the UI that it needs to render the new state. You can use whatever tool/library as long as it satisfies those requirements.

One example of a tool that we can use to achieve the above in a plain view model without depending on molecule is by using [Flow.combine](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/combine.html) + [stateIn](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/state-in.html) operator but this has some drawbacks:

1. You'll need to have your sources as a `Flow` to pass them to combine.
2. This approach scales non-linearly as the number of your sources grow.
3. even with a small number of sources (say 2 or 3), the logic you need to write in the `combine` body will grow out quickly.
4. Last but not least, testing something like this is tricky (add to this the `viewmodelScope` that you will need to fake/substitute in tests)

So, here is where molecule comes to the rescue, it allows you to break down your presenter logic into small composable functions, encapsulate & reuse logic as much as you can, and it makes testing your presenter as simple as testing any function (as you can see here in [PodcastDetailsPresenterTest](https://github.com/mr3y-the-programmer/Podcaster/blob/main/app/src/test/kotlin/com/mr3y/podcaster/ui/presenter/podcastdetails/PodcastDetailsPresenterTest.kt)

So, In summary: Using molecule here is about making the code more idiomatic, maintainable, and more importantly testable.

If you want to learn more about molecule, you can check out the talks about it from previous years conferences and here is a really good [talk](https://youtu.be/XWk_FJm5fEU?si=dXdIqEbPiS-auysW) about it from this year KotlinConf.

Hope this answers your question, I'll close this issue since it is not an actual bug/missing feature in the project itself. Feel free to ask any other question in the discussions.
